src/sentry/eventstream/kafka/state.py

```diff
@@ -0,0 +1,178 @@
+from __future__ import absolute_import
+
+import logging
+import threading
+from collections import defaultdict, namedtuple
+
+
+logger = logging.getLogger(__name__)
+
+
+Offsets = namedtuple('Offsets', 'local remote')
+
+
+class InvalidState(Exception):
+    pass
+
+
+class InvalidStateTransition(Exception):
+    pass
+
+
+class MessageNotReady(Exception):
+    pass
+
+
+class SynchronizedPartitionState:
+    # The ``SYNCHRONIZED`` state represents that the local offset is equal to
+    # the remote offset. The local consumer should be paused to avoid advancing
+    # further beyond the remote consumer.
+    SYNCHRONIZED = 'SYNCHRONIZED'
+
+    # The ``LOCAL_BEHIND`` state represents that the remote offset is greater
+    # than the local offset. The local consumer should be unpaused to avoid
+    # falling behind the remote consumer.
+    LOCAL_BEHIND = 'LOCAL_BEHIND'
+
+    # The ``REMOTE_BEHIND`` state represents that the local offset is greater
+    # than the remote offset. The local consumer should be paused to avoid
+    # advancing further beyond the remote consumer.
+    REMOTE_BEHIND = 'REMOTE_BEHIND'
+
+    # The ``UNKNOWN`` state represents that we haven't received enough data to
+    # know the current offset state.
+    UNKNOWN = 'UNKNOWN'
+
+
+class SynchronizedPartitionStateManager(object):
+    """
+    This class implements a state machine that can be used to track the
+    consumption progress of a Kafka partition (the "local" consumer) relative
+    to a the progress of another consumer (the "remote" consumer.)
+
+    This is intended to be paired with the ``SynchronizedConsumer``.
```

Comment:
General architecture question, does this functionality need to be in its own class, as opposed to just being part of `SynchronizedConsumer` ? I guess it does separate out the kafka-consuming concerns from the state-machine concerns and avoids creating a monster class, but OTOH it doesn't really seem like either one of these classes can really live on its own.

Reply:
Yeah, I don't think they're really required to be independent — I think it's a little easier to understand their responsibilities as separate items (`validate_local_message` is a bit of an aberration here) but I might not have the same perspective as someone less familiar.

Reply:
+1 for separating, just from my memory if this evolving. It's a lot to take in at once otherwise.